File: 2022/day_04.cpp
//
// Created by Wollenschneider Luis on 04.12.22.
//

#include "../utils.hpp"
#include <list>

#define DAY "04"
#define YEAR "2022"

struct Ranges {
    int min1;
    int max1;
    int min2;
    int max2;
};

#define TYPE Ranges

std::list<TYPE> parse_input(const std::string& filepath) {
    // check if path exists
    std::list<TYPE> input;
    std::ifstream file(filepath);
    if (!file.is_open()) {
        std::cout << "File not found" << std::endl;
        return input;
    }
    std::string line;
    while (std::getline(file, line)) {
        unsigned long i = line.find(',');
        std::string range1 = line.substr(0, i);
        std::string range2 = line.substr(i + 1);
        TYPE ranges = {
            .min1 = std::stoi(range1.substr(0, range1.find('-'))),
            .max1 = std::stoi(range1.substr(range1.find('-') + 1)),
            .min2 = std::stoi(range2.substr(0, range2.find('-'))),
            .max2 = std::stoi(range2.substr(range2.find('-') + 1))
        };
        input.push_back(ranges);
    }
    file.close();
    return input;
}

int part1(const std::list<TYPE>& inp) {
    int res = 0;
    for (const TYPE& ranges : inp) {
        if (ranges.min1 >= ranges.min2 && ranges.max1 <= ranges.max2
            || ranges.min2 >= ranges.min1 && ranges.max2 <= ranges.max1)
            res += 1;
    }
    return res;
}

int part2(const std::list<TYPE>& inp) {
    int res = 0;
    for (const TYPE& ranges : inp) {
        if (!(ranges.max1 < ranges.min2 || ranges.max2 < ranges.min1))
            res += 1;
    }
    return res;
}

int main() {
    std::list<TYPE> inp = parse_input(YEAR "/inputs/day_" DAY ".txt");
    std::list<TYPE> test_inp = parse_input(YEAR "/tests/day_" DAY ".txt");

    int expected_result_part1 = get_expected_result<int>(YEAR "/tests/results/day_" DAY "_1.txt");

    int res = part1(test_inp);
    bool eval_res = evaluate_results(res, expected_result_part1);
    res = part1(inp);
    std::cout << "Part 1: " << res << std::endl;

    if (!eval_res) {
        return 1;
    }

    std::cout << std::endl;
    int expected_result_part2 = get_expected_result<int>(YEAR "/tests/results/day_" DAY "_2.txt");

    res = part2(test_inp);
    evaluate_results(res, expected_result_part2);
    res = part2(inp);
    std::cout << "Part 2: " << res << std::endl;

    return 0;
}

File: 2025/day_04.ts
import { solve } from '../utils';

const YEAR = 2025;
const DAY = 4;

function parseInput(input: string): any {
    return input.split('\n').map(line => line.trim().split('').map((c) => c === '@' ? 1 : 0));
}


function neighbors(x: number, y: number, grid: number[][]): number[] {
    let vs = [];
    for (let dx of [-1, 0, 1]) {
        for (let dy of [-1, 0, 1]) {
            if (dx === 0 && dy === 0) continue;
            let nx = x + dx;
            let ny = y + dy;
            if (nx >= 0 && ny >= 0 && ny < grid.length && nx < grid[ny].length) {
                vs.push(grid[ny][nx]);
            }
        }
    }
    return vs;
}

function accessable(grid: number[][]): [number, number][] {
    let as: [number, number][] = [];
    for (let y = 0; y < grid.length; y++) {
        for (let x = 0; x < grid[y].length; x++) {
            if (grid[y][x] === 0) continue;
            if (neighbors(x, y, grid).reduce((a, b) => a + b, 0) < 4) {
                as.push([x, y]);
            }
        }
    }
    return as;
}


function part1(input: string): any | null {
    let inp = parseInput(input);

    return accessable(inp).length;
}

function part2(input: string): any | null {
    let inp = parseInput(input);

    let s = 0;
    while (true) {
        let acc = accessable(inp);
        if (acc.length === 0) {
            break;
        }
        s += acc.length;
        for (let [x, y] of acc) {
            inp[y][x] = 0;
        }
    }

    return s;
}

await solve(part1, YEAR, DAY, 1);
await solve(part2, YEAR, DAY, 2);